gdnative-core/src/globalscope.rs

```diff
@@ -0,0 +1,417 @@
+//! `fmod` is `%` operator on `f32`;
+//! ```rust
+//! use std::ops::Rem;
+//! assert_eq!(f32::rem(12.0, 10.0), 2.0);
+//! ```
+
+use std::f32::consts::TAU;
+use std::ops::RangeInclusive;
+use std::ops::Rem;
+
+const CMP_EPSILON: f32 = 0.00001;
+
+/// Converts a 2D point expressed in the `cartesian` coordinate system (X and Y axis)
+///  to the `polar` coordinate system (a distance from the origin and an angle `(in radians)`).
+#[inline]
+pub fn cartasian2polar(x: f32, y: f32) -> (f32, f32) {
+    ((x * x + y * y).sqrt(), y.atan2(x))
+}
+
+/// Converts from `decibels` to linear energy (audio).
+#[inline]
+pub fn db2linear(db: f32) -> f32 {
+    f32::exp(db * 0.115_129_255)
+}
+
+/// Returns the `position` of the first `non-zero` digit, after the decimal point.
+/// Note that the `maximum` return `value` is `10`, which is a design decision in the implementation.
+/// # Examples:
+/// ```
+/// use gdnative_core::globalscope::step_decimals;
+/// assert_eq!(step_decimals(5.0), 0);
+/// assert_eq!(step_decimals(12.0004), 4);
+/// assert_eq!(step_decimals(0.000000004), 9);
+/// ```
+#[inline]
+pub fn step_decimals(step: f32) -> i32 {
+    const MAXN: usize = 10;
+    const SD: [f32; MAXN] = [
+        0.9999, // somehow compensate for floating point error
+        0.09999,
+        0.009999,
+        0.0009999,
+        0.00009999,
+        0.000009999,
+        0.0000009999,
+        0.00000009999,
+        0.000000009999,
+        0.0000000009999,
+    ];
+
+    let abs = step.abs();
+    let int_abs: i32 = step as i32;
+    let decs: f32 = abs - (int_abs as f32); // strip away integer part;
+    for (i, item) in SD.iter().enumerate().take(MAXN) {
+        if decs >= *item {
+            return i.try_into().unwrap();
+        }
+    }
+    0
+}
+
+/// Moves 'range.start()' toward 'range.end()' by the `delta` `value`.
+/// Use a negative `delta` value `range.end()` move away.
+/// # Examples:
+/// ```
+/// use gdnative_core::globalscope::move_toward;
+/// assert_eq!(move_toward(10.0..=5.0, 4.), 6.);
+/// assert_eq!(move_toward(10.0..=5.0, -1.5), 11.5);
+/// ```
+#[inline]
+pub fn move_toward(range: RangeInclusive<f32>, delta: f32) -> f32 {
```

Comment:
I really like the use of the `RangeInclusive<f32>` for move_towards and other linear interpolation functions. I think that the comments that you provided should help allay any confusion from people migrating from GDScript.

Reply:
sure, will do that